### ISSUE TITLE
React 18 Migration: The Umbrella

### ISSUE THREAD

**Issue by joshblack** [carbon-design-system/carbon#11308]
React shipped their v18 major release earlier this month. We would like to include support for this version in the `carbon-components-react`, `@carbon/react`, and `@carbon/icons-react` packages in v11. In order to do so, we will need to:

```[tasklist]
### Tasks
- [x] https://github.com/carbon-design-system/carbon/issues/10184
- [x] #13220
```

---

### Workarounds

In the meantime, you can force the install of React 18 by either: 
* Install using `--legacy-peer-deps`
* Manually update the dependency version in your project's `node_modules`
* Use yarn's `resolutions` to override the dependency version
* Use npm's `overrides` to override the dependency version
  * ```
    "dependencies": {
      "react": "^18",
      "react-dom": "^18",
    },
    "overrides": {
      "@carbon/react": {
        "react": "^18",
        "react-dom": "^18"
      }
    }
    ```

These workarounds are not a statement of support - there may be compatibility issues with React 18.

**Comment by tay1orjones** [comment]
>  Find an appropriate way in CI to test behavior in v17 and v18

There's some interesting ideas [in this article](https://medium.com/welldone-software/two-ways-to-run-tests-on-different-versions-of-the-same-library-f-e-react-17-react-16-afb7f861d1e9) with using yarn aliases and installing both versions of react. Toggling between the two versions via an env variable seems reasonable

**Comment by dglozic** [comment]
When you DO add support for React 18, will you also add it to v10? Many teams want to move to React 18 (or have already moved using `--legacy-peer-deps`) but are not ready to move to Carbon 11 just for this.

**Comment by sstrubberg** [comment]
hey @dglozic, we have no plans to backport React 18 support to Carbon 10. We are, however, creating time and space for dealing with the fallout from moving to React 18. We expect a spike to investigate this further. Come join us in v11! The water is fine ;)

**Comment by dglozic** [comment]
Based on what I have seen so far, a significant number of products are still on v10 because of the hardship of porting to 11. I am expecting this to become a problem as a significant portion of code running in production is not on a version you are currently spending most of your time on.

React 18 is a good example - we moved from 16.12 (skipping 17) to 18.2. I had to fix couple of things here and there (mostly hydration) and mostly because I wanted to take advantage of `renderToPipeableStream` which required some porting from `renderToNodeStream`. I wish porting to Carbon 11 was the same experience. I think there will be more pressure to add more legacy modes to lower the cost of porting (until it is as easy as React 16 -> 18).

**Comment by The-Real-Thisas** [comment]
Not to be 'that guy', but is there an ETA on this since Next.js 13 uses react 18 by default. I've been using the `--legacy-peer-deps` as a workaround and haven't really had much issues but I don't know if it would have any issues when deploying to like vercel.

**Comment by tay1orjones** [comment]
@The-Real-Thisas We've got a draft PR up that we're actively working on: https://github.com/carbon-design-system/carbon/pull/13332 we're trying to land it asap. We just need to fix the CI issues.

**Comment by tay1orjones** [comment]
Just to update here, React 18 support is out and included starting in v1.27.0

https://github.com/carbon-design-system/carbon/pull/13332

The remaining items outlined in this issue are just pieces that were previously blocked by lack of React 18 support.

**Comment by tay1orjones** [comment]
All items that remained in this issue have been transferred to https://github.com/carbon-design-system/carbon/issues/13623
